main/lsp/TimeTravelingGlobalState.cc

```diff
@@ -270,6 +274,39 @@ bool TimeTravelingGlobalState::canTakeFastPath(const LSPFileUpdates &updates) co
     return true;
 }
 
+LSPFileUpdates TimeTravelingGlobalState::getCombinedUpdates(u4 fromId, u4 toId) {
+    auto ttus = updatesBetweenExclusive(fromId - 1, toId + 1);
+    // Apply backwards so later updates take precedence.
+    reverse(ttus.begin(), ttus.end());
+
+    UnorderedSet<string> encountered;
+    LSPFileUpdates merged;
+    merged.versionEnd = toId;
+    merged.versionStart = fromId;
+    for (const auto ttu : ttus) {
+        const auto &update = ttu->update;
+        int i = -1;
+        merged.hasNewFiles = merged.hasNewFiles || ttu->hasNewFiles;
+        for (const auto &file : update.fileUpdates) {
+            i++;
+            if (!encountered.contains(file->path())) {
+                encountered.insert(string(file->path()));
+                merged.updatedFiles.push_back(file);
+                merged.updatedFileHashes.push_back(update.hashUpdates[i]);
+            }
+        }
+    }
+    merged.canTakeFastPath = canTakeFastPath(fromId - 1, merged);
+
+    // TODO: Avoid re-indexing.
+    for (auto &file : merged.updatedFiles) {
+        auto fref = gs->findFileByPath(file->path());
+        merged.updatedFileIndexes.push_back(pipeline::indexOne(config.opts, *gs, fref, kvstore));
+    }
```

Comment:
What is this section working around?

Reply:
Good question. We cache a set of parsed ASTs for all files in the project, and we depend on them being in sync with the current state of the world. If you change file `foo.rb` and we have to re-typecheck `bar.rb`, we can just copy the AST of `bar.rb` without re-parsing it.

So, we're sending to the typechecking thread the updated indices for the changed files. We *already* sent the typechecking thread indices for these changes (alongside the update that we're trying to cancel), but they are `unique_ptr`s and we no longer have access to them. So, as a hack, I'm recreating them here.

In the future, it might be cleaner to make them `shared_ptr`s -- especially if it's possible to make them read-only and thus shareable among threads -- so we never have to do extra work like this. We always copy them before using them for typechecking.